File: android/app/src/main/java/bh/pray/app/PrayerNotificationSchedule.java
package bh.pray.app;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.Comparator;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.TimeZone;

public final class PrayerNotificationSchedule {
    public static final String IDENTIFIER_PREFIX = "prayer-notification.";
    public static final int MAX_DAYS = 12;
    public static final int MAX_OCCURRENCES = 60;
    private static final TimeZone BAHRAIN = TimeZone.getTimeZone("Asia/Bahrain");

    public interface PreferenceSource {
        boolean isEnabled(String prayer);
        int offsetMinutes(String prayer);
    }

    public static final class Occurrence {
        public final String identifier;
        public final String prayer;
        public final long prayerDateMillis;
        public final long fireMillis;
        public final String prayerTime;

        Occurrence(
            String identifier,
            String prayer,
            long prayerDateMillis,
            long fireMillis,
            String prayerTime
        ) {
            this.identifier = identifier;
            this.prayer = prayer;
            this.prayerDateMillis = prayerDateMillis;
            this.fireMillis = fireMillis;
            this.prayerTime = prayerTime;
        }
    }

    private PrayerNotificationSchedule() {}

    public static List<Occurrence> generate(
        Calendar now,
        int days,
        PreferenceSource preferences
    ) {
        long nowMillis = now.getTimeInMillis();
        Calendar date = Calendar.getInstance(BAHRAIN, Locale.US);
        date.setTimeInMillis(nowMillis);
        date.set(Calendar.HOUR_OF_DAY, 0);
        date.set(Calendar.MINUTE, 0);
        date.set(Calendar.SECOND, 0);
        date.set(Calendar.MILLISECOND, 0);

        List<Occurrence> occurrences = new ArrayList<>();
        int dayCount = Math.max(0, Math.min(days, MAX_DAYS));
        for (int day = 0; day < dayCount; day++) {
            int year = date.get(Calendar.YEAR);
            int month = date.get(Calendar.MONTH) + 1;
            int dayOfMonth = date.get(Calendar.DAY_OF_MONTH);
            Map<String, String> times = PrayerCalculator.calculate(year, month, dayOfMonth);

            for (String prayer : PrayerNotificationPreferences.PRAYER_KEYS) {
                if (!preferences.isEnabled(prayer)) {
                    continue;
                }

                String prayerTime = times.get(prayer);
                Calendar prayerDate = prayerDate(date, prayerTime);
                long prayerMillis = prayerDate.getTimeInMillis();
                long fireMillis = prayerMillis -
                    Math.max(0, preferences.offsetMinutes(prayer)) * 60_000L;
                if (fireMillis <= nowMillis) {
                    continue;
                }

                String identifier = String.format(
                    Locale.US,
                    "%s%04d-%02d-%02d.%s",
                    IDENTIFIER_PREFIX,
                    year,
                    month,
                    dayOfMonth,
                    prayer
                );
                occurrences.add(new Occurrence(
                    identifier,
                    prayer,
                    prayerMillis,
                    fireMillis,
                    prayerTime
                ));
            }
            date.add(Calendar.DAY_OF_MONTH, 1);
        }

        occurrences.sort(Comparator.comparingLong(item -> item.fireMillis));
        if (occurrences.size() > MAX_OCCURRENCES) {
            return new ArrayList<>(occurrences.subList(0, MAX_OCCURRENCES));
        }
        return occurrences;
    }

    private static Calendar prayerDate(Calendar date, String time) {
        String[] parts = time.split(":");
        Calendar result = Calendar.getInstance(BAHRAIN, Locale.US);
        result.clear();
        result.set(
            date.get(Calendar.YEAR),
            date.get(Calendar.MONTH),
            date.get(Calendar.DAY_OF_MONTH),
            Integer.parseInt(parts[0]),
            Integer.parseInt(parts[1]),
            0
        );
        return result;
    }
}

File: ios/PrayBHApp/PrayerNotificationSchedule.swift
import Foundation

struct PrayerNotificationOccurrence: Equatable {
    let identifier: String
    let prayer: NotificationPrayer
    let prayerDate: Date
    let fireDate: Date
    let prayerTime: String
}

enum PrayerNotificationSchedule {
    static let maximumDays = 12
    static let maximumOccurrences = 60
    static let identifierPrefix = "prayer-notification."

    private static var bahrainCalendar: Calendar {
        var calendar = Calendar(identifier: .gregorian)
        calendar.locale = Locale(identifier: "en_US_POSIX")
        calendar.timeZone = TimeZone(identifier: bahrainTimeZoneIdentifier)
            ?? TimeZone(secondsFromGMT: 3 * 3600)!
        return calendar
    }

    static func occurrences(
        startingAt: Date,
        days: Int = maximumDays,
        preferences: PrayerNotificationPreferencesSnapshot,
        maximumCount: Int = maximumOccurrences
    ) -> [PrayerNotificationOccurrence] {
        let calendar = bahrainCalendar
        let startOfToday = calendar.startOfDay(for: startingAt)
        let dayCount = min(max(0, days), maximumDays)
        var result: [PrayerNotificationOccurrence] = []

        for dayOffset in 0..<dayCount {
            guard let date = calendar.date(
                byAdding: .day,
                value: dayOffset,
                to: startOfToday
            ) else {
                continue
            }

            let dateString = PrayerTimesLocal.bahrainDateString(from: date)
            let response = PrayerTimesLocal.buildResponse(dateString: dateString)
            let prayersByKey = Dictionary(
                uniqueKeysWithValues: response.prayers.map { ($0.key, $0) }
            )

            for notificationPrayer in NotificationPrayer.all
            where preferences.isEnabled(notificationPrayer) {
                guard
                    let prayer = prayersByKey[notificationPrayer.rawValue],
                    let prayerDate = makeDate(
                        dateString: dateString,
                        timeString: prayer.time,
                        calendar: calendar
                    ),
                    let fireDate = calendar.date(
                        byAdding: .minute,
                        value: -preferences.offsetMinutes(for: notificationPrayer),
                        to: prayerDate
                    ),
                    fireDate > startingAt
                else {
                    continue
                }

                result.append(
                    PrayerNotificationOccurrence(
                        identifier: identifier(
                            dateString: dateString,
                            prayer: notificationPrayer
                        ),
                        prayer: notificationPrayer,
                        prayerDate: prayerDate,
                        fireDate: fireDate,
                        prayerTime: prayer.time
                    )
                )
            }
        }

        return Array(
            result
                .sorted {
                    if $0.fireDate == $1.fireDate {
                        return $0.identifier < $1.identifier
                    }
                    return $0.fireDate < $1.fireDate
                }
                .prefix(min(max(0, maximumCount), maximumOccurrences))
        )
    }

    private static func makeDate(
        dateString: String,
        timeString: String,
        calendar: Calendar
    ) -> Date? {
        let dateParts = dateString.split(separator: "-").compactMap { Int($0) }
        let timeParts = timeString.split(separator: ":").compactMap { Int($0) }
        guard dateParts.count == 3, timeParts.count == 2 else { return nil }

        var components = DateComponents()
        components.calendar = calendar
        components.timeZone = calendar.timeZone
        components.year = dateParts[0]
        components.month = dateParts[1]
        components.day = dateParts[2]
        components.hour = timeParts[0]
        components.minute = timeParts[1]
        components.second = 0
        return calendar.date(from: components)
    }

    private static func identifier(
        dateString: String,
        prayer: NotificationPrayer
    ) -> String {
        "\(identifierPrefix)\(dateString).\(prayer.rawValue)"
    }
}
